File: main_ws/src/perception/qr_detector_cpp/src/qr_detector_node.cpp
#include "qr_detector_cpp/qr_detector_node.hpp"

#include <ament_index_cpp/get_package_share_directory.hpp>
#include <filesystem>
#include <rclcpp_components/register_node_macro.hpp>

namespace qr_detector_cpp
{

QrDetectorNode::QrDetectorNode(const rclcpp::NodeOptions &options) : Node("qr_detector", options)
{
  // ---- Parameters ----------------------------------------------------
  this->declare_parameter("model_dir", "");
  this->declare_parameter("publish_compressed", true);
  this->declare_parameter("jpeg_quality", 60);
  this->declare_parameter("detection_interval", 1);

  model_dir_ = this->get_parameter("model_dir").as_string();
  publish_compressed_ = this->get_parameter("publish_compressed").as_bool();
  jpeg_quality_ = this->get_parameter("jpeg_quality").as_int();
  detection_interval_ = this->get_parameter("detection_interval").as_int();

  // Resolve model directory
  if (model_dir_.empty())
  {
    try
    {
      std::string python_pkg_share = ament_index_cpp::get_package_share_directory("qr_detector");
      model_dir_ = (std::filesystem::path(python_pkg_share) / "models").string();
    }
    catch (const std::exception &e)
    {
      RCLCPP_WARN(this->get_logger(), "Could not find 'qr_detector' share directory: %s", e.what());
    }
  }

  // ---- WeChatQRCode detector -----------------------------------------
  std::string detect_proto = (std::filesystem::path(model_dir_) / "detect.prototxt").string();
  std::string detect_model = (std::filesystem::path(model_dir_) / "detect.caffemodel").string();
  std::string sr_proto = (std::filesystem::path(model_dir_) / "sr.prototxt").string();
  std::string sr_model = (std::filesystem::path(model_dir_) / "sr.caffemodel").string();

  bool all_files_exist = true;
  for (const auto &f : {detect_proto, detect_model, sr_proto, sr_model})
  {
    if (!std::filesystem::exists(f))
    {
      RCLCPP_WARN(this->get_logger(), "Model file not found: %s", f.c_str());
      all_files_exist = false;
    }
  }

  if (all_files_exist)
  {
    try
    {
      detector_ = std::make_unique<cv::wechat_qrcode::WeChatQRCode>(detect_proto, detect_model,
                                                                    sr_proto, sr_model);
      RCLCPP_INFO(this->get_logger(), "WeChatQRCode detector initialized");
    }
    catch (const std::exception &e)
    {
      RCLCPP_ERROR(this->get_logger(), "Failed to initialize WeChatQRCode: %s", e.what());
    }
  }
  else
  {
    RCLCPP_WARN(this->get_logger(), "WeChatQRCode models not found — using zbar fallback");
  }

  // zbar: QRコードのみ有効化
  zbar_scanner_.set_config(zbar::ZBAR_NONE, zbar::ZBAR_CFG_ENABLE, 0);
  zbar_scanner_.set_config(zbar::ZBAR_QRCODE, zbar::ZBAR_CFG_ENABLE, 1);

  // ---- QoS: sensor data = BEST_EFFORT --------------------------------
  auto sensor_qos = rclcpp::QoS(rclcpp::KeepLast(5))
                        .reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT)
                        .durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);

  // ---- Subscriber ----------------------------------------------------
  image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
      "image_raw", sensor_qos,
      std::bind(&QrDetectorNode::image_callback, this, std::placeholders::_1));

  // ---- Publishers ----------------------------------------------------
  qr_pub_ = this->create_publisher<std_msgs::msg::String>("qr_codes", 10);

  if (publish_compressed_)
  {
    compressed_pub_ =
        this->create_publisher<sensor_msgs::msg::CompressedImage>("image/compressed", sensor_qos);
  }

  RCLCPP_INFO(this->get_logger(), "qr_detector started (interval=%d, jpeg_q=%d, compressed=%d)",
              detection_interval_, jpeg_quality_, publish_compressed_);
}

void QrDetectorNode::image_callback(const sensor_msgs::msg::Image::ConstSharedPtr msg)
{
  frame_count_++;

  if (frame_count_ % detection_interval_ != 0)
  {
    return;
  }

  cv_bridge::CvImagePtr cv_ptr;
  try
  {
    cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
  }
  catch (const cv_bridge::Exception &e)
  {
    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
    return;
  }

  cv::Mat frame = cv_ptr->image;
  std::vector<cv::Mat> points;
  std::vector<std::string> results;

  bool use_zbar = !detector_;

  if (detector_)
  {
    try
    {
      results = detector_->detectAndDecode(frame, points);
    }
    catch (const cv::Exception &e)
    {
      RCLCPP_WARN(this->get_logger(), "WeChatQRCode failed, falling back to zbar: %s", e.what());
      use_zbar = true;
    }
    catch (const std::exception &e)
    {
      RCLCPP_WARN(this->get_logger(), "WeChatQRCode failed, falling back to zbar: %s", e.what());
      use_zbar = true;
    }
    catch (...)
    {
      RCLCPP_WARN(this->get_logger(),
                  "WeChatQRCode failed with unknown error, falling back to zbar");
      use_zbar = true;
    }
  }

  if (use_zbar)
  {
    // zbar フォールバック (MONO8 で直接変換)
    cv_bridge::CvImagePtr gray_ptr;
    try
    {
      gray_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::MONO8);
    }
    catch (const cv_bridge::Exception &e)
    {
      RCLCPP_ERROR(this->get_logger(), "cv_bridge mono8 exception: %s", e.what());
      return;
    }
    cv::Mat gray = gray_ptr->image.isContinuous() ? gray_ptr->image : gray_ptr->image.clone();
    zbar::Image zbar_img(gray.cols, gray.rows, "Y800", gray.data,
                         static_cast<unsigned long>(gray.cols * gray.rows));
    zbar_scanner_.scan(zbar_img);
    for (auto sym = zbar_img.symbol_begin(); sym != zbar_img.symbol_end(); ++sym)
    {
      if (sym->get_type() == zbar::ZBAR_QRCODE)
      {
        results.push_back(sym->get_data());
      }
    }
  }

  for (size_t i = 0; i < results.size(); ++i)
  {
    if (results[i].empty())
    {
      continue;
    }

    auto qr_msg = std_msgs::msg::String();
    qr_msg.data = results[i];
    qr_pub_->publish(qr_msg);
    RCLCPP_INFO(this->get_logger(), "QR detected: %s", results[i].c_str());

    if (publish_compressed_ && i < points.size())
    {
      cv::Mat pts_mat = points[i];
      std::vector<cv::Point> pts;
      for (int row = 0; row < pts_mat.rows; ++row)
      {
        pts.push_back(cv::Point(static_cast<int>(pts_mat.at<float>(row, 0)),
                                static_cast<int>(pts_mat.at<float>(row, 1))));
      }
      cv::polylines(frame, pts, true, cv::Scalar(0, 255, 0), 2);
      cv::putText(frame, results[i].substr(0, 40), cv::Point(pts[0].x, pts[0].y - 10),
                  cv::FONT_HERSHEY_SIMPLEX, 0.6, cv::Scalar(0, 0, 255), 2);
    }
  }

  if (publish_compressed_)
  {
    auto comp_msg = sensor_msgs::msg::CompressedImage();
    comp_msg.header = msg->header;
    comp_msg.format = "jpeg";
    std::vector<int> params = {cv::IMWRITE_JPEG_QUALITY, jpeg_quality_};
    cv::imencode(".jpg", frame, comp_msg.data, params);
    compressed_pub_->publish(comp_msg);
  }
}

}  // namespace qr_detector_cpp

RCLCPP_COMPONENTS_REGISTER_NODE(qr_detector_cpp::QrDetectorNode)

File: main_ws/src/perception/qr_detector/qr_detector/qr_detector_node.py
"""
qr_detector_node: ROS 2 node for real-time QR code detection.

Uses OpenCV's WeChatQRCode (contrib) detector for robust detection.
Optionally publishes JPEG-compressed images for remote monitoring.

Subscribe: /camera/image_raw  (sensor_msgs/Image)
Publish:   /qr_codes          (std_msgs/String)       — detected QR content
Publish:   /image/compressed   (sensor_msgs/CompressedImage) — optional
"""

import os

import cv2
import numpy as np
import rclpy
from ament_index_python.packages import get_package_share_directory
from cv_bridge import CvBridge
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import CompressedImage, Image
from std_msgs.msg import String


class QrDetectorNode(Node):
    def __init__(self) -> None:
        super().__init__("qr_detector")

        # ---- Parameters ----------------------------------------------------
        self.declare_parameter("model_dir", "")
        self.declare_parameter("publish_compressed", True)
        self.declare_parameter("jpeg_quality", 60)
        self.declare_parameter("detection_interval", 1)  # process every N-th frame

        model_dir = self.get_parameter("model_dir").get_parameter_value().string_value
        self._publish_compressed: bool = (
            self.get_parameter("publish_compressed").get_parameter_value().bool_value
        )
        self._jpeg_quality: int = (
            self.get_parameter("jpeg_quality").get_parameter_value().integer_value
        )
        self._detection_interval: int = (
            self.get_parameter("detection_interval").get_parameter_value().integer_value
        )

        # Resolve model directory
        if not model_dir:
            # Try installed share directory first, fall back to source
            try:
                pkg_share = get_package_share_directory("qr_detector")
                model_dir = os.path.join(pkg_share, "models")
            except Exception:
                model_dir = os.path.join(
                    os.path.dirname(__file__), "..", "..", "models"
                )

        # ---- WeChatQRCode detector -----------------------------------------
        detect_proto = os.path.join(model_dir, "detect.prototxt")
        detect_model = os.path.join(model_dir, "detect.caffemodel")
        sr_proto = os.path.join(model_dir, "sr.prototxt")
        sr_model = os.path.join(model_dir, "sr.caffemodel")

        for f in (detect_proto, detect_model, sr_proto, sr_model):
            if not os.path.isfile(f):
                self.get_logger().warn(f"Model file not found: {f}")

        self._detector = cv2.wechat_qrcode_WeChatQRCode(
            detect_proto, detect_model, sr_proto, sr_model
        )
        self.get_logger().info("WeChatQRCode detector initialized")

        # ---- cv_bridge -----------------------------------------------------
        self._bridge = CvBridge()

        # ---- QoS: sensor data = BEST_EFFORT --------------------------------
        sensor_qos = QoSProfile(
            depth=5,
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
        )

        # ---- Subscriber ----------------------------------------------------
        self.create_subscription(
            Image, "/camera/image_raw", self._image_callback, sensor_qos
        )

        # ---- Publishers ----------------------------------------------------
        self._qr_pub = self.create_publisher(String, "/qr_codes", 10)

        if self._publish_compressed:
            self._compressed_pub = self.create_publisher(
                CompressedImage, "/image/compressed", sensor_qos
            )
        else:
            self._compressed_pub = None

        self._frame_count: int = 0

        self.get_logger().info(
            f"qr_detector started (interval={self._detection_interval}, "
            f"jpeg_q={self._jpeg_quality}, compressed={self._publish_compressed})"
        )

    # -----------------------------------------------------------------------
    def _image_callback(self, msg: Image) -> None:
        self._frame_count += 1

        # Frame skipping for performance
        if self._frame_count % self._detection_interval != 0:
            return

        try:
            frame = self._bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
        except Exception as e:
            self.get_logger().error(f"cv_bridge error: {e}")
            return

        # ---- QR detection --------------------------------------------------
        results, points = self._detector.detectAndDecode(frame)

        for i, content in enumerate(results):
            if not content:
                continue

            qr_msg = String()
            qr_msg.data = content
            self._qr_pub.publish(qr_msg)
            self.get_logger().info(f"QR detected: {content}")

            # Draw bounding box on frame for compressed output
            if (
                self._compressed_pub is not None
                and points is not None
                and i < len(points)
            ):
                pts = points[i].astype(np.int32).reshape(-1, 2)
                cv2.polylines(
                    frame, [pts], isClosed=True, color=(0, 255, 0), thickness=2
                )
                cv2.putText(
                    frame,
                    content[:40],
                    (pts[0][0], pts[0][1] - 10),
                    cv2.FONT_HERSHEY_SIMPLEX,
                    0.6,
                    (0, 0, 255),
                    2,
                )

        # ---- Publish compressed image (with annotations) -------------------
        if self._compressed_pub is not None:
            comp_msg = CompressedImage()
            comp_msg.header = msg.header
            comp_msg.format = "jpeg"
            _, buf = cv2.imencode(
                ".jpg", frame, [cv2.IMWRITE_JPEG_QUALITY, self._jpeg_quality]
            )
            comp_msg.data = buf.tobytes()
            self._compressed_pub.publish(comp_msg)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = QrDetectorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
